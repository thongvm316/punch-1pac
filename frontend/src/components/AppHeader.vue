<template>
  <header class="header">
    <div class="container grid-xl">
      <div class="navbar">
        <section class="navbar-section">
          <router-link to="/dashboard" class="navbar-brand mr-4">
            <img :src="companyLogoUrl" :alt="companyName">
          </router-link>
          <router-link class="mr-5" to="/dashboard">{{ $t('header.dashboard') }}</router-link>
          <router-link class="mr-5" to="/attendances">{{ $t('header.attendances') }}</router-link>
          <router-link class="mr-5" to="/requests">{{ $t('header.requests') }}</router-link>
          <router-link class="mr-5" to="/groups">{{ $t('header.groups') }}</router-link>
        </section>
        <section class="navbar-section">
          <punch/>
          <header-announcements/>
          <div class="dropdown" :class="{ active: isDropdownActive }" @click="toggleDropdown">
            <a class="dropdown-toggle">
              <div class="tile tile-centered">
                <div class="tile-icon">
                  <img :src="userAvatarUrl" class="avatar avatar-md" :alt="userName">
                </div>
                <div class="tile-content mr-2">{{ userName }}</div>
              </div>
            </a>
            <ul class="menu">
              <li class="menu-item"><router-link to="/settings">{{ $t('header.settings') }}</router-link></li>
              <li class="menu-item"><router-link to="/company/settings">{{ $t('header.companySettings') }}</router-link></li>
              <li class="menu-item"><a href="#" @click="logout($event)">{{ $t('header.logout') }}</a></li>
            </ul>
          </div>
        </section>
      </div>
    </div>
  </header>
</template>

<script>
import axios from 'axios'
import HeaderAnnouncements from '../components/HeaderAnnouncements.vue'
import Punch from '../components/Punch.vue'

export default {
  name: 'app-header',
  data () {
    return {
      isDropdownActive: false
    }
  },

  components: {
    HeaderAnnouncements,
    Punch
  },

  methods: {
    logout () {
      axios.post('/logout').then(() => { window.location.href = '/' })
    },

    toggleDropdown () {
      this.isDropdownActive = !this.isDropdownActive
    }
  },

  computed: {
    companyName () {
      return window.INITIAL_STATE.company.name
    },

    companyLogoUrl () {
      return window.INITIAL_STATE.company.logo_url
    },

    userName () {
      return window.INITIAL_STATE.user.name
    },

    userAvatarUrl () {
      return window.INITIAL_STATE.user.avatar_url
    }
  }
}
</script>
