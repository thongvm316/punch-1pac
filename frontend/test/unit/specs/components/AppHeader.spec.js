import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'
import initialStatesData from '../../supports/fixtures/initial-states.api'
import Repositories from '@/repository'
import AppHeader from '@/components/AppHeader'
jest.mock('@/repository/users')

const usersRepository = Repositories.get('users')
const logout = jest.spyOn(AppHeader.methods, 'logout')
const toggleDropdown = jest.spyOn(AppHeader.methods, 'toggleDropdown')
const toggleLangSelect = jest.spyOn(AppHeader.methods, 'toggleLangSelect')
const updateUser = jest.spyOn(AppHeader.methods, 'updateUser')
const INITIAL_STATES_UPDATE_USER_LANGUAGE = jest.spyOn(AppHeader.methods, 'INITIAL_STATES_UPDATE_USER_LANGUAGE')

Object.assign(wrapperOps, {
  Notifications: true,
  AnnualLeave: true,
  Punch: true
})

describe('AppHeader.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(AppHeader, wrapperOps)
    setComputed(wrapper, {
      meta: initialStatesData.meta
    })
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when app header was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when toggleDropdown', () => {
    it('should called toggleDropdown method', async () => {
      wrapper.find({ ref: 'dropdownMenu' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(toggleDropdown).toHaveBeenCalled()
      expect(wrapper.vm.isDropdownActive).toEqual(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when toggleLanguageSelect', () => {
    it('should called toogleLangSelect method', async () => {
      wrapper.find({ ref: 'toggleLangSelectBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(toggleLangSelect).toHaveBeenCalled()
      expect(wrapper.vm.isLangSelectActive).toEqual(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when logout method', () => {
    it('should call logout method', async () => {
      wrapper.find({ ref: 'logoutBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(logout).toHaveBeenCalled()
    })
  })

  describe('when user change language', () => {
    it('should call updateUser method', async () => {
      usersRepository.updateUser.mockResolvedValue(null)
      wrapper.setData({ isLangSelectActive: true })
      wrapper.find({ ref: 'toggleLangSelectBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      wrapper.findAll({ ref: 'languageItem' }).at(2).find('a').trigger('click')

      expect(updateUser).toHaveBeenCalled()
    })
  })
})
