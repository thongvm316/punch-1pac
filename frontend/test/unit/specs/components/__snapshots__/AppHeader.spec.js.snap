// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader.vue when app header was mounted should render correctly 1`] = `
<header class="header">
  <div class="container">
    <div class="navbar">
      <section class="navbar-section">
        <router-link-stub to="/dashboard" tag="a" event="click" class="navbar-brand mr-4"><img></router-link-stub>
        <router-link-stub to="/attendances" tag="a" event="click" class="mr-5">
          Attendances
        </router-link-stub>
        <router-link-stub to="/requests" tag="a" event="click" class="mr-5">
          Requests
        </router-link-stub>
        <!---->
      </section>
      <section class="navbar-section">
        <div>
          <div class="punch">
            <!---->
            <!----> <span class="mr-5">10:39:37</span> <button class="btn btn-primary mr-5">
              Punch In
            </button>
            <!---->
          </div>
          <!---->
        </div>
        <div><button type="button" class="btn btn-notice mr-5">
            Day Off
          </button>
          <!---->
        </div>
        <div>
          <div class="dropdown mr-5"><span data-badge="0" class="notification"><p-ico-bell-stub></p-ico-bell-stub></span>
            <div class="box notifications notification-dropdown triangle-top">
              <div class="notification-header">
                <h4>Notifications</h4>
              </div>
              <p class="no-notification-msg">
                You have no new notifications
              </p>
            </div>
          </div>
          <!---->
        </div>
        <div class="dropdown"><a class="dropdown-toggle">
            <div class="tile tile-centered mr-2">
              <div class="tile-icon"><img class="avatar avatar-md"></div>
            </div>
          </a>
          <ul class="menu triangle-top">
            <li class="menu-item">
              <router-link-stub to="/settings" tag="a" event="click">
                Settings
              </router-link-stub>
            </li>
            <li class="menu-item">
              <!---->
            </li>
            <li class="menu-item"><a>Change Language...</a></li>
            <li class="menu-item"><a href="#">Logout</a></li>
          </ul>
          <ul class="menu triangle-top lang-select" style="display: none;">
            <li class="header"><a class="btn-back"></a>
              <p>Languages</p>
            </li>
            <li class="menu-item"><a class="active">English</a></li>
            <li class="menu-item"><a class="">Vietnamese</a></li>
            <li class="menu-item"><a class="">Japanese</a></li>
          </ul>
        </div>
      </section>
    </div>
  </div>
</header>
`;

exports[`AppHeader.vue when toggleDropdown should called toggleDropdown method 1`] = `
<header class="header">
  <div class="container">
    <div class="navbar">
      <section class="navbar-section">
        <router-link-stub to="/dashboard" tag="a" event="click" class="navbar-brand mr-4"><img></router-link-stub>
        <router-link-stub to="/attendances" tag="a" event="click" class="mr-5">
          Attendances
        </router-link-stub>
        <router-link-stub to="/requests" tag="a" event="click" class="mr-5">
          Requests
        </router-link-stub>
        <!---->
      </section>
      <section class="navbar-section">
        <div>
          <div class="punch">
            <!---->
            <!----> <span class="mr-5">10:39:38</span> <button class="btn btn-primary mr-5">
              Punch In
            </button>
            <!---->
          </div>
          <div class="modal"><span class="modal-overlay"></span>
            <div class="modal-container">
              <div class="modal-header"><span class="btn btn-clear float-right"></span>
                <h3 class="modal-title">
                  Punch Out
                </h3>
              </div>
              <div class="modal-body">
                <div class="content">
                  <p>Do you intent to punch out at 10:39 ?</p>
                </div>
              </div>
              <div class="modal-footer"><button type="button" class="btn btn-error">
                  Yes
                </button> <button type="button" class="btn">
                  No
                </button></div>
            </div>
          </div>
        </div>
        <div><button type="button" class="btn btn-notice mr-5">
            Day Off
          </button>
          <div class="modal"><span class="modal-overlay"></span>
            <div class="modal-container">
              <div class="modal-header"><span class="btn btn-clear float-right"></span>
                <h3 class="modal-title">
                  Request a leave
                </h3>
              </div>
              <div class="modal-body">
                <div class="content"></div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="dropdown mr-5"><span data-badge="0" class="notification"><p-ico-bell-stub></p-ico-bell-stub></span>
            <div class="box notifications notification-dropdown triangle-top">
              <div class="notification-header">
                <h4>Notifications</h4>
              </div>
              <p class="no-notification-msg">
                You have no new notifications
              </p>
            </div>
          </div>
          <!---->
        </div>
        <div class="dropdown active"><a class="dropdown-toggle">
            <div class="tile tile-centered mr-2">
              <div class="tile-icon"><img class="avatar avatar-md"></div>
            </div>
          </a>
          <ul class="menu triangle-top">
            <li class="menu-item">
              <router-link-stub to="/settings" tag="a" event="click">
                Settings
              </router-link-stub>
            </li>
            <li class="menu-item">
              <!---->
            </li>
            <li class="menu-item"><a>Change Language...</a></li>
            <li class="menu-item"><a href="#">Logout</a></li>
          </ul>
          <ul class="menu triangle-top lang-select" style="display: none;">
            <li class="header"><a class="btn-back"></a>
              <p>Languages</p>
            </li>
            <li class="menu-item"><a class="active">English</a></li>
            <li class="menu-item"><a class="">Vietnamese</a></li>
            <li class="menu-item"><a class="">Japanese</a></li>
          </ul>
        </div>
      </section>
    </div>
  </div>
</header>
`;

exports[`AppHeader.vue when toggleLanguageSelect should called toogleLangSelect method 1`] = `
<header class="header">
  <div class="container">
    <div class="navbar">
      <section class="navbar-section">
        <router-link-stub to="/dashboard" tag="a" event="click" class="navbar-brand mr-4"><img></router-link-stub>
        <router-link-stub to="/attendances" tag="a" event="click" class="mr-5">
          Attendances
        </router-link-stub>
        <router-link-stub to="/requests" tag="a" event="click" class="mr-5">
          Requests
        </router-link-stub>
        <!---->
      </section>
      <section class="navbar-section">
        <div>
          <div class="punch">
            <!---->
            <!----> <span class="mr-5">10:39:38</span> <button class="btn btn-primary mr-5">
              Punch In
            </button>
            <!---->
          </div>
          <div class="modal"><span class="modal-overlay"></span>
            <div class="modal-container">
              <div class="modal-header"><span class="btn btn-clear float-right"></span>
                <h3 class="modal-title">
                  Punch Out
                </h3>
              </div>
              <div class="modal-body">
                <div class="content">
                  <p>Do you intent to punch out at 10:39 ?</p>
                </div>
              </div>
              <div class="modal-footer"><button type="button" class="btn btn-error">
                  Yes
                </button> <button type="button" class="btn">
                  No
                </button></div>
            </div>
          </div>
        </div>
        <div><button type="button" class="btn btn-notice mr-5">
            Day Off
          </button>
          <div class="modal"><span class="modal-overlay"></span>
            <div class="modal-container">
              <div class="modal-header"><span class="btn btn-clear float-right"></span>
                <h3 class="modal-title">
                  Request a leave
                </h3>
              </div>
              <div class="modal-body">
                <div class="content"></div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <div class="dropdown mr-5"><span data-badge="0" class="notification"><p-ico-bell-stub></p-ico-bell-stub></span>
            <div class="box notifications notification-dropdown triangle-top">
              <div class="notification-header">
                <h4>Notifications</h4>
              </div>
              <p class="no-notification-msg">
                You have no new notifications
              </p>
            </div>
          </div>
          <!---->
        </div>
        <div class="dropdown"><a class="dropdown-toggle">
            <div class="tile tile-centered mr-2">
              <div class="tile-icon"><img class="avatar avatar-md"></div>
            </div>
          </a>
          <ul class="menu triangle-top" style="display: none;">
            <li class="menu-item">
              <router-link-stub to="/settings" tag="a" event="click">
                Settings
              </router-link-stub>
            </li>
            <li class="menu-item">
              <!---->
            </li>
            <li class="menu-item"><a>Change Language...</a></li>
            <li class="menu-item"><a href="#">Logout</a></li>
          </ul>
          <ul class="menu triangle-top lang-select" style="">
            <li class="header"><a class="btn-back"></a>
              <p>Languages</p>
            </li>
            <li class="menu-item"><a class="active">English</a></li>
            <li class="menu-item"><a class="">Vietnamese</a></li>
            <li class="menu-item"><a class="">Japanese</a></li>
          </ul>
        </div>
      </section>
    </div>
  </div>
</header>
`;
